src/hotspot/share/runtime/deoptimization.cpp

```diff
@@ -1525,6 +1531,69 @@ void Deoptimization::load_class_by_index(const constantPoolHandle& constant_pool
   }
 }
 
+#if INCLUDE_JFR
+
+class DeoptReasonSerializer : public JfrSerializer {
+ public:
+  void serialize(JfrCheckpointWriter& writer) {
+    writer.write_count((u4)(Deoptimization::Reason_LIMIT + 1)); // + Reason::many (-1)
+    for (int i = -1; i < Deoptimization::Reason_LIMIT; ++i) {
+      writer.write_key((u8)i);
+      writer.write(Deoptimization::trap_reason_name(i));
+    }
+  }
+};
+
+class DeoptActionSerializer : public JfrSerializer {
+ public:
+  void serialize(JfrCheckpointWriter& writer) {
+    static const u4 nof_actions = Deoptimization::Action_LIMIT;
+    writer.write_count(nof_actions);
+    for (u4 i = 0; i < Deoptimization::Action_LIMIT; ++i) {
+      writer.write_key(i);
+      writer.write(Deoptimization::trap_action_name((int)i));
+    }
+  }
+};
+
+static void register_serializers() {
+  static int critical_section = 0;
+  if (1 == critical_section || Atomic::cmpxchg(1, &critical_section, 0) == 1) {
```

Comment:
In the original PR in jdk11 the ordering of the arguments in `Atomic::cmpxchg` have been refactored to a different order. I placed the arguments to the correct order for jdk11. The jdk14 ordering can be found [here](https://github.com/openjdk/jdk/blob/e9f97b2e8cf301ba6b69101e5efc5c71d26bc87b/src/hotspot/share/runtime/atomic.hpp#L719). And the jdk11 ordering can be found [here](https://github.com/microsoft/openjdk-jdk11u/blob/2f695bbaad2620507e8427610ae988332aaa001c/src/hotspot/share/runtime/atomic.hpp#L693).

Reply:
Thanks for providing the links to atomic.hpp!